src/AbcSmc.cpp

```diff
@@ -37,279 +40,408 @@ bool file_exists(const char *fileName) {
+    const string short_name = mpar.get("short_name", name).asString();
+
+    const string ptype_str = mpar["dist_type"].asString();
+    const string ntype_str = mpar["num_type"].asString();
+
+    ABC::Parameter *par;
+
+    if (not ((ntype_str == "INT") or (ntype_str == "FLOAT"))) {
+        cerr << "Unknown parameter numeric type: " << ntype_str << ".  Aborting." << endl;
+        exit(-206);
+    }
+
+    if (ptype_str == "UNIFORM") {
+        if (ntype_str == "INT") {
+            const long par1 = mpar["par1"].asInt64();
+            const long par2 = mpar["par2"].asInt64();
+            par = new DiscreteUniformPrior(name, short_name, par1, par2);
+        } else {
+            const float_type par1 = mpar["par1"].asDouble();
+            const float_type par2 = mpar["par2"].asDouble();
+            par = new ContinuousUniformPrior(name, short_name, par1, par2);
         }
-        const size_t max_set = max(_predictive_prior_sizes.size(), _smc_set_sizes.size());
-        for (size_t i = 0; i < max_set; ++i) {
-            if (get_pred_prior_size(i, false) > get_num_particles(i, false)) {
-                cerr << "Error: requested predictive prior size is greater than requested SMC set size for at least one set in configuration file." << endl;
-                exit(1);
+    } else if (ptype_str == "NORMAL" or ptype_str == "GAUSSIAN") {
+        if (ntype_str == "INT") {
+            cerr << "Parameter numeric " << ntype_str << " not supported for parameter type " << ptype_str << ".  Aborting." << endl;
+            exit(-206);
+        }
+        const float_type par1 = mpar["par1"].asDouble();
+        const float_type par2 = mpar["par2"].asDouble();
+        par = new GaussianPrior(name, short_name, par1, par2);
+    } else if (ptype_str == "PSEUDO") {
+        std::vector<float_type> states;
+        if (mpar.isMember("vals")) {
+            states = as_vector<float_type>(mpar["vals"]);
+        } else {
+            const float_type smax = mpar["par2"].asDouble();
+            const float_type step = mpar.get("step", 1.0).asDouble();
+            if (step != 0) {
+                for (float_type s = mpar["par1"].asDouble(); s <= smax; s += step) {
+                    states.push_back(s);
+                }
+            } else {
+                states.push_back(mpar["par1"].asDouble());
             }
         }
+        par = new PseudoPar(name, short_name, states);
+    } else if (ptype_str == "POSTERIOR") {
+        // TODO iota + stride?
+        const size_t maxIdx = mpar["par2"].asUInt64();
+        par = new PosteriorPar(name, short_name, maxIdx);
+    } else {
+        cerr << "Unknown parameter distribution type: " << ptype_str << ".  Aborting." << endl;
+        exit(-205);
     }
+    return par;
 }
 
-
-bool AbcSmc::parse_config(string conf_filename) {
-    if (not file_exists(conf_filename.c_str())) {
-        cerr << "File does not exist: " << conf_filename << endl;
+Json::Value prepare(const string & configfile) {
```

Comment:
because "file" can reasonably be a filename or a filehandle, I prefer to be explicit about this